`etcdctl`: Migrate `etcd-diagnosis` as a new subcommand `diagnosis`

seshachalam-yv opened the pull request:
Fixes https://github.com/etcd-io/etcd/issues/20275
Part of https://github.com/etcd-io/etcd/issues/20217

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

etcdctl sample output
```bash
✗ ./bin/etcdctl diagnosis --cluster -o diagnosis.json
```

cc @ahrtr

ahrtr commented on etcdctl/ctlv3/command/diagnosis_command.go:
```diff
@@ -0,0 +1,72 @@
+
+	cmd.Flags().BoolVar(&diagCfg.PrintVersion, "version", false, "print the version and exit")
+
+	cmd.Flags().BoolVar(&diagCfg.Offline, "offline", false, "offline analysis")
```
Please do not backport https://github.com/ahrtr/etcd-diagnosis/pull/63. We can migrate it to `etcdutl` separtely.

ahrtr commented on etcdctl/ctlv3/command/diagnosis_command.go:
```diff
@@ -0,0 +1,72 @@
+		Run:   runDiagnosis,
+	}
+
+	cmd.Flags().StringSliceVar(&diagCfg.Endpoints, "endpoints", []string{"127.0.0.1:2379"}, "comma separated etcd endpoints")
+	cmd.Flags().BoolVar(&diagCfg.UseClusterEndpoints, "cluster", false, "use all endpoints from the cluster member list")
+
+	cmd.Flags().DurationVar(&diagCfg.DialTimeout, "dial-timeout", 2*time.Second, "dial timeout for client connections")
+	cmd.Flags().DurationVar(&diagCfg.CommandTimeout, "command-timeout", 5*time.Second, "command timeout (excluding dial timeout)")
+	cmd.Flags().DurationVar(&diagCfg.KeepAliveTime, "keepalive-time", 2*time.Second, "keepalive time for client connections")
+	cmd.Flags().DurationVar(&diagCfg.KeepAliveTimeout, "keepalive-timeout", 5*time.Second, "keepalive timeout for client connections")
+
+	cmd.Flags().BoolVar(&diagCfg.Insecure, "insecure-transport", true, "disable transport security for client connections")
+	cmd.Flags().BoolVar(&diagCfg.InsecureSkipVerify, "insecure-skip-tls-verify", false, "skip server certificate verification (CAUTION: this option should be enabled only for testing purposes)")
+	cmd.Flags().StringVar(&diagCfg.CertFile, "cert", "", "identify secure client using this TLS certificate file")
+	cmd.Flags().StringVar(&diagCfg.KeyFile, "key", "", "identify secure client using this TLS key file")
+	cmd.Flags().StringVar(&diagCfg.CaFile, "cacert", "", "verify certificates of TLS-enabled secure servers using this CA bundle")
+
+	cmd.Flags().StringVar(&diagCfg.Username, "user", "", "username[:password] for authentication (prompt if password is not supplied)")
+	cmd.Flags().StringVar(&diagCfg.Password, "password", "", "password for authentication (if this option is used, --user option shouldn't include password)")
+	cmd.Flags().StringVarP(&diagCfg.DNSDomain, "discovery-srv", "d", "", "domain name to query for SRV records describing cluster endpoints")
+	cmd.Flags().StringVarP(&diagCfg.DNSService, "discovery-srv-name", "", "", "service name to query when using DNS discovery")
+	cmd.Flags().BoolVar(&diagCfg.InsecureDiscovery, "insecure-discovery", true, "accept insecure SRV records describing cluster endpoints")
```
etcdctl already has these flags in ctl.go.

seshachalam-yv replied:
Ack. Removed duplicated flags; diagnosis now reuses the existing `etcdctl` global flags. Only subcommand-specific flags remain.

ahrtr commented on etcdctl/diagnosis/engine/diagnosis.go:
```diff
@@ -0,0 +1,47 @@
+)
+
+const (
+	reportFileName = "etcd_diagnosis_report.json"
```
I suggest that we add a flag `--output` (short `-o`) so that users can set the output file name.

seshachalam-yv replied:
Done. Added -o/--output. When provided, the JSON report is written to the given file; otherwise it’s printed to stdout.

seshachalam-yv commented:
Thanks @ahrtr for your review. I’ve addressed your comments.
Apologies for the delay — I was away for an extended period due to personal reasons.

ahrtr commented:
Please license header for each new go file. https://github.com/etcd-io/etcd/blob/636bbddf65056f87828bbfe4e953ac9e5dea4365/tests/e2e/reproduce_18667_test.go#L1-L13

Also I suggest to move the directory diagnosis to `etcdctl/ctlv3/command/diagnosis`

ahrtr commented on etcdctl/README.md:
```diff
@@ -1129,6 +1129,43 @@ DOWNGRADE CANCEL cancels the ongoing downgrade action to cluster.
+#### Flags
+
+- `--cluster`: use all endpoints discovered from the cluster member list.
+- `--etcd-storage-quota-bytes`: expected etcd storage quota in bytes (value passed to etcd with `--quota-backend-bytes`).
```
For etcd version >= 3.6.0, we are able to get the quota by `endpoint status` query. For versions < 3.6.0, we still need to users to provide a value for this flag.

It's OK to enhance this in a followup PR.

ahrtr reviewed (changes requested):
Please resolve the workflow failures, thx

seshachalam-yv commented:
Hi @ahrtr ,

I have updated `etcdctl/README.md` to resolve multiple markdown linting errors (MD003, MD004, MD007, MD012, MD022, MD024). Since I modified `etcdctl/README.md` for the diagnosis docs, it triggered the markdown lint jobs, requiring me to fix the lint errors throughout the file.

**Changes include:**

*   **Standardized Headers**: Converted Setext-style headers (underlined) to ATX-style (`#`) to fix MD003.
*   **Duplicate Headings**: Renamed generic headings like `#### Output` and `#### Example` to be unique (e.g., `#### PUT Example`) to fix MD024.
*   **Spacing**: Removed excessive blank lines (MD012) and fixed spacing around headers (MD022).
*   **Lists**: Corrected list indentation and styles (MD004, MD007).

ahrtr commented on etcdctl/README.md:
```diff
@@ -26,7 +25,7 @@ PUT assigns the specified value with the specified key. If key already holds a v
 RPC: Put
 
-#### Options
+#### PUT options \key\ \value\ Options
```
I suggest that we ignore the linter rule instead of updating this.

seshachalam-yv replied:
Agreed, the lint fixes were creating too much noise. I reverted them and instead used an inline markdownlint-disable comment to suppress the existing errors. The PR should be much cleaner now.
